build_runner/lib/src/generate/build_impl.dart

```diff
@@ -329,14 +339,20 @@ class BuildImpl {
                 .where((id) => !inputNode.primaryOutputs.contains(id))
                 .join(', '));
 
-    if (!_buildShouldRun(builderOutputs)) return <AssetId>[];
-
     var wrappedReader = new SingleStepReader(
         _reader,
         _assetGraph,
         phaseNumber,
         input.package,
         (phase, input) => _runLazyPhaseForInput(phase, input, resourceManager));
+
+    if (!await _buildShouldRun(builderOutputs, wrappedReader)) {
+      return <AssetId>[];
+    }
+    // We may have read some inputs in the call to `_buildShouldRun`, we want
```

Comment:
should we not be going through the wrapped reader if we don't want to track this?

Or is that the only way for the lazy assets to work?